Courses endpoint with pagination

marcospri opened the pull request:
For:

- https://github.com/hypothesis/lms/issues/6393

Expose a API endpoint to list courses with pagination.

The intended use case is to list the courses in a drop down menu or similar to allow for filtering.


### Example response

`http://localhost:8001/api/dashboard/courses?limit=2`

```
{
  "courses": [
    {
      "id": 1883,
      "title": "Copy of Developer Test Course with Sections Enabled (With Pags)"
    },
    {
      "id": 1829,
      "title": "Course copy LTI1.3"
    }
  ],
  "pagination": {
    "next": "http://localhost:8001/api/dashboard/courses?cursor=%5B%22Course+copy+LTI1.3%22%2C+1829%5D&limit=2"
  }
}
```


### Testing

- Make sure you have an authorization cookie from the LMS admin pages visiting http://localhost:8001/admin/
- Open http://localhost:8001/api/dashboard/courses
- Play with the limit parameter, eg: http://localhost:8001/api/dashboard/courses?limit=1
- Follow the `next` parameter in the response.

marcospri commented on lms/views/dashboard/api/course.py:
```diff
@@ -22,16 +63,44 @@ def __init__(self, request) -> None:
+        if cursor := self.request.parsed_params.get("cursor"):
+            cursor_course_name, cursor_course_id = _get_cursor_values(cursor)
+            courses = courses.filter(
+                (Course.lms_name, Course.id) > (cursor_course_name, cursor_course_id)
```
This tuple syntax is equivalent to:

name > name or (name = name and id > id)

marcospri commented on lms/js_config_types.py:
```diff
@@ -40,6 +40,8 @@ class APIStudent(TypedDict):
 class APICourses(TypedDict):
     courses: list[APICourse]
 
+    next: NotRequired[str]
```
Responses now might contain the cursor

marcospri commented on lms/routes.py:
```diff
@@ -258,6 +258,8 @@ def includeme(config):  # noqa: PLR0915
         factory="lms.resources.dashboard.DashboardResource",
     )
 
+    config.add_route("api.dashboard.courses", "/api/dashboard/courses")
```
Now endpoint for listing this. We could in theory reuse one single endpoint and using parameters construct the results, ie:

`?with_assigment_counts=1`

For now I reckon it would be simpler to have one endpoint that returns the results to display in the table an another for listing them for the dropdowns.

marcospri commented on lms/services/course.py:
```diff
@@ -140,19 +140,33 @@ def search(  # noqa: PLR0913, PLR0917
 
-    def get_organization_courses(
-        self, organization: Organization, h_userid: str | None
+    def get_courses(
```
Making organization optional here, changing the name accordingly.

marcospri commented on lms/services/course.py:
```diff
@@ -140,19 +140,33 @@ def search(  # noqa: PLR0913, PLR0917
-            # Deduplicate courses by authority_provided_id, take the last updated one
+
+        # Deduplicate courses by authority_provided_id, take the last updated one
+        deduplicated_courses = select(Course.id).select_from(
```
We need the deduplication step.

marcospri commented on lms/services/course.py:
```diff
@@ -140,19 +140,33 @@ def search(  # noqa: PLR0913, PLR0917
+        return (
+            self._db.query(Course)
+            .filter(Course.id.in_(deduplicated_courses))
+            .order_by(Course.lms_name, Course.id)
```
Order by the column we'll paginate with. Maybe in the future this could be an arbitrary set of columns.

Let's not try to build an arbitrary pagination _system_ with the first paginated endpoint.

marcospri replied:
We need to columns to make for a more stable order than just the name. This could eventually become:

 Course.name, Course.**term**, Course.id but I'd always include the ID there for good measure at the end.

robertknight replied:
> We need to columns to make for a more stable order than just the name. 

You mean _two_ columns, right?

marcospri replied:
Yes :facepalm:

marcospri commented on lms/views/dashboard/api/course.py:
```diff
@@ -8,12 +12,49 @@
+class ListCoursesSchema(PyramidRequestSchema):
+    location = "query"
+
+    organization = fields.Str()
```
TODO, actually use this to filter the results.

Just as an example here that this endpoint (and the equivalent ones in assignments, users) will need filtering on top of pagination.

marcospri commented on lms/views/dashboard/api/course.py:
```diff
@@ -8,12 +12,49 @@
+
+    cursor_data = (last_element.lms_name, last_element.id)
+    return (
+        base64.urlsafe_b64encode(json.dumps(cursor_data).encode("utf-8"))
```
base64-ing a json list is probably the most immediate solution to encode two values.

marcospri commented on lms/views/dashboard/api/course.py:
```diff
@@ -8,12 +12,49 @@
+    return (
+        base64.urlsafe_b64encode(json.dumps(cursor_data).encode("utf-8"))
+        .decode("utf-8")
+        .replace("=", "")
```
Remove the padding, it looks ugly in URLs.

marcospri commented on lms/views/dashboard/api/course.py:
```diff
@@ -8,12 +12,49 @@
+
+def _get_cursor_values(coursor: str) -> tuple[str, int]:
+    """Decode the values from a next cursor."""
+    return json.loads(base64.urlsafe_b64decode(coursor + "==="))
```
Add back the padding.

marcospri commented on lms/views/dashboard/api/course.py:
```diff
@@ -22,16 +63,44 @@ def __init__(self, request) -> None:
+        )
+
+        limit = min(PAGINATION_LIMIT, self.request.parsed_params["limit"])
+        if cursor := self.request.parsed_params.get("cursor"):
```
We might not have a cursor, for example for the first set of results.

marcospri commented on lms/views/dashboard/api/course.py:
```diff
@@ -22,16 +63,44 @@ def __init__(self, request) -> None:
+            "courses": [
+                APICourse(id=course.id, title=course.lms_name) for course in courses
+            ],
+            "next": _next_parameter(courses),
```
`next` here at the top level result, should we include it on a "pagination" object or similar?

marcospri commented on lms/views/dashboard/api/course.py:
```diff
@@ -30,8 +98,9 @@ def __init__(self, request) -> None:
         org = get_request_organization(self.request, self.organization_service)
-        courses = self.course_service.get_organization_courses(
-            org, h_userid=self.request.user.h_userid if self.request.user else None
+        courses = self.course_service.get_courses(
```
Account for the signature change on this method on the other view.

robertknight reviewed (commented):
One design question I have is whether `next` should be a complete URL rather than just a value for a query param. I think that might be slightly easier to use for clients and perhaps allow the server to change how pagination is represented in the URL without needing to update the frontend in future.

marcospri commented on lms/views/dashboard/api/course.py:
```diff
@@ -7,13 +11,71 @@
+"""Maximum number of items to return in paginated endpoints"""
+
+
+def build_courses_pagination(
```
This has a lot of room for generalization and abstraction.

I'd like to resit the urge to do that until the second endpoint (assignments) and then refine on the third (students).

robertknight replied:
I think a name like `get_courses_page` would make it more obvious that the endpoint returns one page of results plus the pagination metadata. This could then generalize to something like `get_page(request, items: list[T], limit) -> tuple[list[T], Pagination]` in future.

marcospri commented on tests/unit/lms/services/course_test.py:
```diff
@@ -366,7 +366,7 @@ def test_get_members(self, svc, db_session):
         ) == [user]
 
-    def test_get_organization_courses_deduplicates(self, db_session, svc):
+    def test_get_courses_deduplicates(self, db_session, svc):
```
Account for the service name changes

robertknight commented on lms/views/dashboard/api/course.py:
```diff
@@ -7,13 +11,71 @@
+    """Maximum number of items to return."""
+
+    cursor = fields.Str()
+    """Position to return the elements from. This correponds to the `next` value in the repsonse."""
```
```suggestion
    """Position to return elements from."""
```

robertknight commented on lms/views/dashboard/api/course.py:
```diff
@@ -7,13 +11,71 @@
+            return in_data
+
+        try:
+            in_data["cursor"] = json.loads(base64.urlsafe_b64decode(cursor + "==="))
```
As noted above, the base64 encoding doesn't seem strictly necessary to me.

robertknight commented on lms/services/course.py:
```diff
@@ -140,19 +140,33 @@ def search(  # noqa: PLR0913, PLR0917
+        return (
+            self._db.query(Course)
+            .filter(Course.id.in_(courses_query))
+            # We can sort these again without affecting deduplication
```
To check my understanding, the reason for using two queries here rather than doing the search with just one, is to enable using one sort order for the inner query, and then a different sort order for the outer query?

robertknight commented on lms/views/dashboard/api/course.py:
```diff
@@ -7,13 +11,71 @@
+def build_courses_pagination(
+    request, courses, limit: int = PAGINATION_LIMIT
+) -> tuple[list[Course], Pagination]:
+    """Build the pagination details for a list of courses.
```
It wasn't immediately clear here that `list[Course]` is a single page of results. A docstring like "Return the first page and pagination metadata from a list of courses" would make that clear.

robertknight commented on lms/js_config_types.py:
```diff
@@ -7,6 +7,10 @@
 
 
+class Pagination(TypedDict):
+    next: str | None
```
A docstring would be useful here to make it clear this is a URL.

robertknight reviewed (commented):
Functionally this works as expected. I left a few notes on various details. Also there are some references to the earlier API design for pagination in the comments.

marcospri commented on lms/views/dashboard/api/course.py:
```diff
@@ -7,13 +11,71 @@
+    )
+
+
+class ListCoursesSchema(PyramidRequestSchema):
```
My first instinct would be to make this: `_ReponseWithPaginationMixin` or similar.

`ListCoursesSchema` will just subclass it (and clarify it doesn't need extra query parameters)

`ListAssignemtnsSchema` will subclass it and include for example `course_id` as an optional filter.

marcospri commented:
Removed the base64-ing of the cursor

Also added the right type annotation to the method that build the pagination parameters. It doesn't takte a list of courses but rather a queryset (just Query[Course] in SQLA)
